The terraform-provider-azurerm_v4.21.1_x5.exe plugin crashed

### Is there an existing issue for this?

- [x] I have searched the existing issues

### Community Note

<!--- Please keep this note for the community --->

* Please vote on this issue by adding a :thumbsup: [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave comments along the lines of "+1", "me too" or "any updates", they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment and review the [contribution guide](https://github.com/hashicorp/terraform-provider-azurerm/blob/main/contributing/README.md) to help.

<!--- Thank you for keeping this note for the community --->


### Terraform Version

1.11.0

### AzureRM Provider Version

4.21.1

### Affected Resource(s)/Data Source(s)

azurerm_mssql_server

### Terraform Configuration Files

```hcl
https://github.com/golfalot/terraform-bug-report-01
```

### Debug Output/Panic Output

```shell
(base) PS D:\source\repos-vscode\Infrastructure-nonprod\nonprod-transacting> terraform init -upgrade
Initializing the backend...
Initializing provider plugins...
- Finding hashicorp/azurerm versions matching "~> 4.19"...
- Finding latest version of hashicorp/random...
- Finding latest version of hashicorp/null...
- Installing hashicorp/azurerm v4.21.1...
- Installed hashicorp/azurerm v4.21.1 (signed by HashiCorp)
- Using previously-installed hashicorp/random v3.7.1
- Using previously-installed hashicorp/null v3.2.3
Terraform has made some changes to the provider dependency selections recorded
in the .terraform.lock.hcl file. Review those changes and commit them to your
version control system if they represent changes you intended to make.

Terraform has been successfully initialized!

You may now begin working with Terraform. Try running "terraform plan" to see
any changes that are required for your infrastructure. All Terraform commands
should now work.

If you ever set or change modules or backend configuration for Terraform,
rerun this command to reinitialize your working directory. If you forget, other
commands will detect it and remind you to do so if necessary.
(base) PS D:\source\repos-vscode\Infrastructure-nonprod\nonprod-transacting> terraform plan
data.azurerm_subscription.current: Reading...

Terraform used the selected providers to generate the following execution plan. Resource actions are indicated with the following symbols:
  + create

Terraform planned the following actions, but then encountered a problem:

  # azurerm_resource_group.rg_sql will be created
  + resource "azurerm_resource_group" "rg_sql" {
      + id       = (known after apply)
      + location = "uksouth"
      + name     = "rg-sql-transacting-nonprod-uksouth"
    }

  # random_password.sql_admin_password will be created
  + resource "random_password" "sql_admin_password" {
      + bcrypt_hash      = (sensitive value)
      + id               = (known after apply)
      + length           = 16
      + lower            = true
      + min_lower        = 0
      + min_numeric      = 0
      + min_special      = 0
      + min_upper        = 0
      + number           = true
      + numeric          = true
      + override_special = "_%@"
      + result           = (sensitive value)
      + special          = true
      + upper            = true
    }

  # random_string.sql_username_suffix will be created
  + resource "random_string" "sql_username_suffix" {
      + id          = (known after apply)
      + length      = 8
      + lower       = true
      + min_lower   = 0
      + min_numeric = 0
      + min_special = 0
      + min_upper   = 0
      + number      = true
      + numeric     = true
      + result      = (known after apply)
      + special     = false
      + upper       = false
    }

Plan: 3 to add, 0 to change, 0 to destroy.
╷
│ Error: Plugin did not respond
│
│   with data.azurerm_subscription.current,
│   on providers.tf line 20, in data "azurerm_subscription" "current":
│   20: data "azurerm_subscription" "current" {}
│
│ The plugin encountered an error, and failed to respond to the plugin.(*GRPCProvider).ReadDataSource call. The plugin logs may contain more details.
╵
╷
│ Error: Plugin did not respond
│
│   with azurerm_mssql_server.sql,
│   on sql_server.tf line 6, in resource "azurerm_mssql_server" "sql":
│    6: resource "azurerm_mssql_server" "sql" {
│
│ The plugin encountered an error, and failed to respond to the plugin.(*GRPCProvider).PlanResourceChange call. The plugin logs may contain more details.
╵

Stack trace from the terraform-provider-azurerm_v4.21.1_x5.exe plugin:

panic: value is unknown

goroutine 77 [running]:
github.com/hashicorp/go-cty/cty.Value.AsString({{{0x8b3ee68?, 0xc00000a821?}}, {0x72a24a0?, 0xdcb9be0?}})
        github.com/hashicorp/go-cty@v1.4.1-0.20200414143053-d3edf31b6320/cty/value_ops.go:1179 +0xf8
github.com/hashicorp/terraform-provider-azurerm/internal/services/mssql.msSqlAdministratorLoginPassword({0x0?, 0x0?}, 0xc003088e00, {0x0?, 0x0?})
        github.com/hashicorp/terraform-provider-azurerm/internal/services/mssql/mssql_server_resource.go:766 +0x9b
github.com/hashicorp/terraform-provider-azurerm/internal/services/mssql.resourceMsSqlServer.CustomizeDiffShim.func8({0x8b3a168?, 0xc00308edb0?}, 0x0?, {0x729cb00?, 0xc00200a908?})
        github.com/hashicorp/terraform-provider-azurerm/internal/tf/pluginsdk/shim.go:18 +0x2c
github.com/hashicorp/terraform-provider-azurerm/internal/services/mssql.resourceMsSqlServer.CustomDiffWithAll.func9({0x8b3a168, 0xc00308edb0}, 0xc003088e00, {0x729cb00, 0xc00200a908})
        github.com/hashicorp/terraform-provider-azurerm/internal/tf/pluginsdk/customize_diff.go:30 +0xab
github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema.schemaMap.Diff(0xc00150b290, {0x8b3a168, 0xc00308edb0}, 0xc001710f70, 0xc00180a410, 0xc001219020, {0x729cb00, 0xc00200a908}, 0x0)
        github.com/hashicorp/terraform-plugin-sdk/v2@v2.36.0/helper/schema/schema.go:710 +0x4b4
github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema.(*Resource).SimpleDiff(0x8b3dc78?, {0x8b3a168?, 0xc00308edb0?}, 0xc001710f70, 0xc00308ede0?, {0x729cb00?, 0xc00200a908?})
        github.com/hashicorp/terraform-plugin-sdk/v2@v2.36.0/helper/schema/resource.go:1022 +0xdb
github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema.(*GRPCProviderServer).PlanResourceChange(0xc000c08108, {0x8b3a168?, 0xc00308ecc0?}, 0xc00180a230)
        github.com/hashicorp/terraform-plugin-sdk/v2@v2.36.0/helper/schema/grpc_provider.go:890 +0xbce
github.com/hashicorp/terraform-plugin-mux/tf5muxserver.(*muxServer).PlanResourceChange(0xc0002a0000, {0x8b3a168?, 0xc00308e9f0?}, 0xc00180a230)
        github.com/hashicorp/terraform-plugin-mux@v0.17.0/tf5muxserver/mux_server_PlanResourceChange.go:73 +0x2ad
github.com/hashicorp/terraform-plugin-go/tfprotov5/tf5server.(*server).PlanResourceChange(0xc001a641e0, {0x8b3a168?, 0xc00308e180?}, 0xc00309c080)
        github.com/hashicorp/terraform-plugin-go@v0.26.0/tfprotov5/tf5server/server.go:826 +0x3f0
github.com/hashicorp/terraform-plugin-go/tfprotov5/internal/tfplugin5._Provider_PlanResourceChange_Handler({0x808cc80, 0xc001a641e0}, {0x8b3a168, 0xc00308e180}, 0xc00309c000, 0x0)
        github.com/hashicorp/terraform-plugin-go@v0.26.0/tfprotov5/internal/tfplugin5/tfplugin5_grpc.pb.go:593 +0x1a6
google.golang.org/grpc.(*Server).processUnaryRPC(0xc0007ea200, {0x8b3a168, 0xc00308e120}, 0xc002d800c0, 0xc001ed3530, 0xdc22a80, 0x0)
        google.golang.org/grpc@v1.69.4/server.go:1392 +0xfd2
google.golang.org/grpc.(*Server).handleStream(0xc0007ea200, {0x8b3de70, 0xc001c96b60}, 0xc002d800c0)
        google.golang.org/grpc@v1.69.4/server.go:1802 +0xbaa
google.golang.org/grpc.(*Server).serveStreams.func2.1()
        google.golang.org/grpc@v1.69.4/server.go:1030 +0x8b
created by google.golang.org/grpc.(*Server).serveStreams.func2 in goroutine 47
        google.golang.org/grpc@v1.69.4/server.go:1041 +0x125

Error: The terraform-provider-azurerm_v4.21.1_x5.exe plugin crashed!

This is always indicative of a bug within the plugin. It would be immensely
helpful if you could report the crash with the plugin's maintainers so that it
can be fixed. The output above should help diagnose the issue.
```

### Expected Behaviour

no crash

### Actual Behaviour

Error: The terraform-provider-azurerm_v4.21.1_x5.exe plugin crashed!

### Steps to Reproduce

1. clone repo
2. set valid subscription id and name in variables.tf
3. terraform init -upgrade
4. terraform plan

### Important Factoids

_No response_

### References

_No response_

--- a/internal/services/mssql/mssql_server_resource.go
+++ b/internal/services/mssql/mssql_server_resource.go
@@ -84,7 +84,7 @@ func resourceMsSqlServer() *pluginsdk.Resource {
 				Computed:     true,
 				ForceNew:     true,
 				AtLeastOneOf: []string{"administrator_login", "azuread_administrator.0.azuread_authentication_only"},
-				ValidateFunc: validation.StringIsNotWhiteSpace,
+				ValidateFunc: validation.StringIsNotEmpty,
 			},
 
 			"administrator_login_password": {

--- a/internal/services/mssql/mssql_server_resource.go
+++ b/internal/services/mssql/mssql_server_resource.go
@@ -84,6 +84,7 @@ func resourceMsSqlServer() *pluginsdk.Resource {
 				Computed:     true,
 				ForceNew:     true,
 				AtLeastOneOf: []string{"administrator_login", "azuread_administrator.0.azuread_authentication_only"},
+				ValidateFunc: validation.StringIsNotWhiteSpace,
 			},
 
 			"administrator_login_password": {
@@ -760,10 +761,10 @@ func msSqlPasswordChangeWhenAADAuthOnly(ctx context.Context, d *pluginsdk.Resour
 }
 
 // msSqlAdministratorLoginPassword checks to make sure that one of `administrator_login_password_wo` or `administrator_login_password` is set when `administrator_login` is specified.
-func msSqlAdministratorLoginPassword(ctx context.Context, d *pluginsdk.ResourceDiff, _ interface{}) (err error) {
+func msSqlAdministratorLoginPassword(_ context.Context, d *pluginsdk.ResourceDiff, _ interface{}) (err error) {
 	adminLogin, ok := d.GetRawConfig().AsValueMap()["administrator_login"]
 	if ok {
-		if !adminLogin.IsNull() && adminLogin.AsString() != "" {
+		if !adminLogin.IsNull() {
 			woAdminLoginPassword, err := pluginsdk.GetWriteOnlyFromDiff(d, "administrator_login_password_wo", cty.String)
 			if err != nil {
 				return err

--- a/internal/services/mssql/mssql_server_resource_test.go
+++ b/internal/services/mssql/mssql_server_resource_test.go
@@ -41,6 +41,21 @@ func TestAccMsSqlServer_basic(t *testing.T) {
 	})
 }
 
+func TestAccMsSqlServer_basicWithUnknownValue(t *testing.T) {
+	data := acceptance.BuildTestData(t, "azurerm_mssql_server", "test")
+	r := MsSqlServerResource{}
+
+	data.ResourceTest(t, r, []acceptance.TestStep{
+		{
+			Config: r.basicWithUnknownValue(data),
+			Check: acceptance.ComposeTestCheckFunc(
+				check.That(data.ResourceName).ExistsInAzure(r),
+			),
+		},
+		data.ImportStep("administrator_login_password"),
+	})
+}
+
 func TestAccMsSqlServer_complete(t *testing.T) {
 	data := acceptance.BuildTestData(t, "azurerm_mssql_server", "test")
 	r := MsSqlServerResource{}
@@ -410,6 +425,23 @@ resource "azurerm_mssql_server" "test" {
 `, r.template(data), data.RandomInteger)
 }
 
+func (r MsSqlServerResource) basicWithUnknownValue(data acceptance.TestData) string {
+	return fmt.Sprintf(`
+%s
+
+resource "azurerm_mssql_server" "test" {
+  name                         = "acctestsqlserver%[2]d"
+  resource_group_name          = azurerm_resource_group.test.name
+  location                     = azurerm_resource_group.test.location
+  version                      = "12.0"
+  administrator_login          = "missadministrator-${random_string.test.result}"
+  administrator_login_password = "thisIsKat11"
+
+  outbound_network_restriction_enabled = true
+}
+`, r.template(data), data.RandomInteger)
+}
+
 func (r MsSqlServerResource) basicWithMinimumTLSVersionDisabled(data acceptance.TestData) string {
 	return fmt.Sprintf(`
 %s
@@ -513,7 +545,7 @@ resource "azurerm_mssql_server" "test" {
   resource_group_name          = azurerm_resource_group.test.name
   location                     = azurerm_resource_group.test.location
   version                      = "12.0"
-  administrator_login          = "missadministrator-${random_string.test.result}"
+  administrator_login          = "missadministrator"
   administrator_login_password = "thisIsKat11"
   minimum_tls_version          = "1.2"
 

--- a/internal/services/mssql/mssql_server_resource_test.go
+++ b/internal/services/mssql/mssql_server_resource_test.go
@@ -318,6 +318,12 @@ func TestAccMsSqlServer_writeOnlyAdminLoginPassword(t *testing.T) {
 			tfversion.SkipBelow(version.Must(version.NewVersion("1.11.0"))),
 		},
 		ProtoV5ProviderFactories: framework.ProtoV5ProviderFactoriesInit(context.Background(), "azurerm"),
+		ExternalProviders: map[string]resource.ExternalProvider{
+			"random": {
+				VersionConstraint: "=3.6.3",
+				Source:            "registry.terraform.io/hashicorp/random",
+			},
+		},
 		Steps: []resource.TestStep{
 			{
 				Config: r.writeOnlyAdminLoginPassword(data, "7h1515K4711-secret", 1),
@@ -342,6 +348,12 @@ func TestAccMsSqlServer_updateToWriteOnlyPassword(t *testing.T) {
 			tfversion.SkipBelow(version.Must(version.NewVersion("1.11.0"))),
 		},
 		ProtoV5ProviderFactories: framework.ProtoV5ProviderFactoriesInit(context.Background(), "azurerm"),
+		ExternalProviders: map[string]resource.ExternalProvider{
+			"random": {
+				VersionConstraint: "=3.6.3",
+				Source:            "registry.terraform.io/hashicorp/random",
+			},
+		},
 		Steps: []resource.TestStep{
 			{
 				Config: r.basic(data),
@@ -381,19 +393,12 @@ func (MsSqlServerResource) Exists(ctx context.Context, client *clients.Client, s
 	return pointer.To(resp.Model != nil), nil
 }
 
-func (MsSqlServerResource) basic(data acceptance.TestData) string {
+func (r MsSqlServerResource) basic(data acceptance.TestData) string {
 	return fmt.Sprintf(`
-provider "azurerm" {
-  features {}
-}
-
-resource "azurerm_resource_group" "test" {
-  name     = "acctestRG-mssql-%[1]d"
-  location = "%[2]s"
-}
+%s
 
 resource "azurerm_mssql_server" "test" {
-  name                         = "acctestsqlserver%[1]d"
+  name                         = "acctestsqlserver%[2]d"
   resource_group_name          = azurerm_resource_group.test.name
   location                     = azurerm_resource_group.test.location
   version                      = "12.0"
@@ -402,22 +407,15 @@ resource "azurerm_mssql_server" "test" {
 
   outbound_network_restriction_enabled = true
 }
-`, data.RandomInteger, data.Locations.Primary)
+`, r.template(data), data.RandomInteger)
 }
 
-func (MsSqlServerResource) basicWithMinimumTLSVersionDisabled(data acceptance.TestData) string {
+func (r MsSqlServerResource) basicWithMinimumTLSVersionDisabled(data acceptance.TestData) string {
 	return fmt.Sprintf(`
-provider "azurerm" {
-  features {}
-}
-
-resource "azurerm_resource_group" "test" {
-  name     = "acctestRG-mssql-%[1]d"
-  location = "%[2]s"
-}
+%s
 
 resource "azurerm_mssql_server" "test" {
-  name                         = "acctestsqlserver%[1]d"
+  name                         = "acctestsqlserver%[2]d"
   resource_group_name          = azurerm_resource_group.test.name
   location                     = azurerm_resource_group.test.location
   version                      = "12.0"
@@ -429,22 +427,15 @@ resource "azurerm_mssql_server" "test" {
     type = "SystemAssigned"
   }
 }
-`, data.RandomInteger, data.Locations.Primary)
+`, r.template(data), data.RandomInteger)
 }
 
-func (MsSqlServerResource) basicWithMinimumTLSVersion(data acceptance.TestData) string {
+func (r MsSqlServerResource) basicWithMinimumTLSVersion(data acceptance.TestData) string {
 	return fmt.Sprintf(`
-provider "azurerm" {
-  features {}
-}
-
-resource "azurerm_resource_group" "test" {
-  name     = "acctestRG-mssql-%[1]d"
-  location = "%[2]s"
-}
+%s
 
 resource "azurerm_mssql_server" "test" {
-  name                         = "acctestsqlserver%[1]d"
+  name                         = "acctestsqlserver%[2]d"
   resource_group_name          = azurerm_resource_group.test.name
   location                     = azurerm_resource_group.test.location
   version                      = "12.0"
@@ -456,7 +447,7 @@ resource "azurerm_mssql_server" "test" {
     type = "SystemAssigned"
   }
 }
-`, data.RandomInteger, data.Locations.Primary)
+`, r.template(data), data.RandomInteger)
 }
 
 func (r MsSqlServerResource) requiresImport(data acceptance.TestData) string {
@@ -474,26 +465,19 @@ resource "azurerm_mssql_server" "import" {
 `, r.basic(data))
 }
 
-func (MsSqlServerResource) complete(data acceptance.TestData) string {
+func (r MsSqlServerResource) complete(data acceptance.TestData) string {
 	return fmt.Sprintf(`
-provider "azurerm" {
-  features {}
-}
-
-resource "azurerm_resource_group" "test" {
-  name     = "acctestRG-mssql-%[1]d"
-  location = "%[2]s"
-}
+%s
 
 resource "azurerm_virtual_network" "test" {
-  name                = "acctestvnet-%[1]d"
+  name                = "acctestvnet-%[2]d"
   resource_group_name = azurerm_resource_group.test.name
   location            = azurerm_resource_group.test.location
   address_space       = ["10.5.0.0/16"]
 }
 
 resource "azurerm_subnet" "service" {
-  name                 = "acctestsnetservice-%[1]d"
+  name                 = "acctestsnetservice-%[2]d"
   resource_group_name  = azurerm_resource_group.test.name
   virtual_network_name = azurerm_virtual_network.test.name
   address_prefixes     = ["10.5.1.0/24"]
@@ -502,7 +486,7 @@ resource "azurerm_subnet" "service" {
 }
 
 resource "azurerm_subnet" "endpoint" {
-  name                 = "acctestsnetendpoint-%[1]d"
+  name                 = "acctestsnetendpoint-%[2]d"
   resource_group_name  = azurerm_resource_group.test.name
   virtual_network_name = azurerm_virtual_network.test.name
   address_prefixes     = ["10.5.2.0/24"]
@@ -525,11 +509,11 @@ resource "azurerm_user_assigned_identity" "test" {
 }
 
 resource "azurerm_mssql_server" "test" {
-  name                         = "acctestsqlserver%[1]d"
+  name                         = "acctestsqlserver%[2]d"
   resource_group_name          = azurerm_resource_group.test.name
   location                     = azurerm_resource_group.test.location
   version                      = "12.0"
-  administrator_login          = "missadministrator"
+  administrator_login          = "missadministrator-${random_string.test.result}"
   administrator_login_password = "thisIsKat11"
   minimum_tls_version          = "1.2"
 
@@ -553,41 +537,34 @@ resource "azurerm_private_dns_zone" "finance" {
 }
 
 resource "azurerm_private_endpoint" "test" {
-  name                = "acctest-privatelink-%[1]d"
+  name                = "acctest-privatelink-%[2]d"
   location            = azurerm_resource_group.test.location
   resource_group_name = azurerm_resource_group.test.name
   subnet_id           = azurerm_subnet.endpoint.id
 
   private_service_connection {
-    name                           = "acctest-privatelink-mssc-%[1]d"
+    name                           = "acctest-privatelink-mssc-%[2]d"
     private_connection_resource_id = azurerm_mssql_server.test.id
     subresource_names              = ["sqlServer"]
     is_manual_connection           = false
   }
 }
-`, data.RandomInteger, data.Locations.Primary, data.RandomIntOfLength(15))
+`, r.template(data), data.RandomInteger, data.RandomIntOfLength(15))
 }
 
-func (MsSqlServerResource) completeUpdate(data acceptance.TestData) string {
+func (r MsSqlServerResource) completeUpdate(data acceptance.TestData) string {
 	return fmt.Sprintf(`
-provider "azurerm" {
-  features {}
-}
-
-resource "azurerm_resource_group" "test" {
-  name     = "acctestRG-mssql-%[1]d"
-  location = "%[2]s"
-}
+%s
 
 resource "azurerm_virtual_network" "test" {
-  name                = "acctestvnet-%[1]d"
+  name                = "acctestvnet-%[2]d"
   resource_group_name = azurerm_resource_group.test.name
   location            = azurerm_resource_group.test.location
   address_space       = ["10.5.0.0/16"]
 }
 
 resource "azurerm_subnet" "service" {
-  name                 = "acctestsnetservice-%[1]d"
+  name                 = "acctestsnetservice-%[2]d"
   resource_group_name  = azurerm_resource_group.test.name
   virtual_network_name = azurerm_virtual_network.test.name
   address_prefixes     = ["10.5.1.0/24"]
@@ -596,7 +573,7 @@ resource "azurerm_subnet" "service" {
 }
 
 resource "azurerm_subnet" "endpoint" {
-  name                 = "acctestsnetendpoint-%[1]d"
+  name                 = "acctestsnetendpoint-%[2]d"
   resource_group_name  = azurerm_resource_group.test.name
   virtual_network_name = azurerm_virtual_network.test.name
   address_prefixes     = ["10.5.2.0/24"]
@@ -619,7 +596,7 @@ resource "azurerm_user_assigned_identity" "test" {
 }
 
 resource "azurerm_mssql_server" "test" {
-  name                         = "acctestsqlserver%[1]d"
+  name                         = "acctestsqlserver%[2]d"
   resource_group_name          = azurerm_resource_group.test.name
   location                     = azurerm_resource_group.test.location
   version                      = "12.0"
@@ -647,34 +624,27 @@ resource "azurerm_private_dns_zone" "finance" {
 }
 
 resource "azurerm_private_endpoint" "test" {
-  name                = "acctest-privatelink-%[1]d"
+  name                = "acctest-privatelink-%[2]d"
   location            = azurerm_resource_group.test.location
   resource_group_name = azurerm_resource_group.test.name
   subnet_id           = azurerm_subnet.endpoint.id
 
   private_service_connection {
-    name                           = "acctest-privatelink-mssc-%[1]d"
+    name                           = "acctest-privatelink-mssc-%[2]d"
     private_connection_resource_id = azurerm_mssql_server.test.id
     subresource_names              = ["sqlServer"]
     is_manual_connection           = false
   }
 }
-`, data.RandomInteger, data.Locations.Primary, data.RandomIntOfLength(15))
+`, r.template(data), data.RandomInteger, data.RandomIntOfLength(15))
 }
 
-func (MsSqlServerResource) systemAssignedIdentity(data acceptance.TestData) string {
+func (r MsSqlServerResource) systemAssignedIdentity(data acceptance.TestData) string {
 	return fmt.Sprintf(`
-provider "azurerm" {
-  features {}
-}
-
-resource "azurerm_resource_group" "test" {
-  name     = "acctestRG-mssql-%[1]d"
-  location = "%[2]s"
-}
+%s
 
 resource "azurerm_mssql_server" "test" {
-  name                         = "acctestsqlserver%[1]d"
+  name                         = "acctestsqlserver%[2]d"
   resource_group_name          = azurerm_resource_group.test.name
   location                     = azurerm_resource_group.test.location
   version                      = "12.0"
@@ -685,19 +655,12 @@ resource "azurerm_mssql_server" "test" {
     type = "SystemAssigned"
   }
 }
-`, data.RandomInteger, data.Locations.Primary)
+`, r.template(data), data.RandomInteger)
 }
 
-func (MsSqlServerResource) userAssignedIdentity(data acceptance.TestData) string {
+func (r MsSqlServerResource) userAssignedIdentity(data acceptance.TestData) string {
 	return fmt.Sprintf(`
-provider "azurerm" {
-  features {}
-}
-
-resource "azurerm_resource_group" "test" {
-  name     = "acctestRG-mssql-%[1]d"
-  location = "%[2]s"
-}
+%s
 
 resource "azurerm_user_assigned_identity" "test1" {
   resource_group_name = azurerm_resource_group.test.name
@@ -712,7 +675,7 @@ resource "azurerm_user_assigned_identity" "test2" {
 }
 
 resource "azurerm_mssql_server" "test" {
-  name                              = "acctestsqlserver%[1]d"
+  name                              = "acctestsqlserver%[2]d"
   resource_group_name               = azurerm_resource_group.test.name
   location                          = azurerm_resource_group.test.location
   version                           = "12.0"
@@ -725,28 +688,21 @@ resource "azurerm_mssql_server" "test" {
     identity_ids = [azurerm_user_assigned_identity.test1.id, azurerm_user_assigned_identity.test2.id]
   }
 }
-`, data.RandomInteger, data.Locations.Primary)
+`, r.template(data), data.RandomInteger)
 }
 
-func (MsSqlServerResource) systemAndUserAssignedIdentity(data acceptance.TestData) string {
+func (r MsSqlServerResource) systemAndUserAssignedIdentity(data acceptance.TestData) string {
 	return fmt.Sprintf(`
-provider "azurerm" {
-  features {}
-}
-
-resource "azurerm_resource_group" "test" {
-  name     = "acctestRG-mssql-%[1]d"
-  location = "%[2]s"
-}
+%s
 
 resource "azurerm_user_assigned_identity" "test" {
-  name                = "acctestUAI-%[1]d"
+  name                = "acctestUAI-%[2]d"
   location            = azurerm_resource_group.test.location
   resource_group_name = azurerm_resource_group.test.name
 }
 
 resource "azurerm_mssql_server" "test" {
-  name                         = "acctestsqlserver%[1]d"
+  name                         = "acctestsqlserver%[2]d"
   resource_group_name          = azurerm_resource_group.test.name
   location                     = azurerm_resource_group.test.location
   version                      = "12.0"
@@ -760,26 +716,19 @@ resource "azurerm_mssql_server" "test" {
     ]
   }
 }
-`, data.RandomInteger, data.Locations.Primary)
+`, r.template(data), data.RandomInteger)
 }
 
-func (MsSqlServerResource) aadAdmin(data acceptance.TestData) string {
+func (r MsSqlServerResource) aadAdmin(data acceptance.TestData) string {
 	return fmt.Sprintf(`
-provider "azurerm" {
-  features {}
-}
+%s
 
 provider "azuread" {}
 
 data "azurerm_client_config" "test" {}
 
-resource "azurerm_resource_group" "test" {
-  name     = "acctestRG-mssql-%[1]d"
-  location = "%[2]s"
-}
-
 resource "azurerm_mssql_server" "test" {
-  name                         = "acctestsqlserver%[1]d"
+  name                         = "acctestsqlserver%[2]d"
   resource_group_name          = azurerm_resource_group.test.name
   location                     = azurerm_resource_group.test.location
   version                      = "12.0"
@@ -791,26 +740,19 @@ resource "azurerm_mssql_server" "test" {
     object_id      = data.azurerm_client_config.test.object_id
   }
 }
-`, data.RandomInteger, data.Locations.Primary)
+`, r.template(data), data.RandomInteger)
 }
 
-func (MsSqlServerResource) aadAdminWithAADAuthOnly(data acceptance.TestData) string {
+func (r MsSqlServerResource) aadAdminWithAADAuthOnly(data acceptance.TestData) string {
 	return fmt.Sprintf(`
-provider "azurerm" {
-  features {}
-}
+%s
 
 provider "azuread" {}
 
 data "azurerm_client_config" "test" {}
 
-resource "azurerm_resource_group" "test" {
-  name     = "acctestRG-mssql-%[1]d"
-  location = "%[2]s"
-}
-
 resource "azurerm_mssql_server" "test" {
-  name                = "acctestsqlserver%[1]d"
+  name                = "acctestsqlserver%[2]d"
   resource_group_name = azurerm_resource_group.test.name
   location            = azurerm_resource_group.test.location
   version             = "12.0"
@@ -821,23 +763,16 @@ resource "azurerm_mssql_server" "test" {
     azuread_authentication_only = true
   }
 }
-`, data.RandomInteger, data.Locations.Primary)
+`, r.template(data), data.RandomInteger)
 }
 
-func (MsSqlServerResource) updateAzureadAuthenticationOnlyWithIdentity(data acceptance.TestData, enableAzureadAuthenticationOnly bool) string {
+func (r MsSqlServerResource) updateAzureadAuthenticationOnlyWithIdentity(data acceptance.TestData, enableAzureadAuthenticationOnly bool) string {
 	return fmt.Sprintf(`
-provider "azurerm" {
-  features {}
-}
+%s
 
 provider "azuread" {}
 
 data "azurerm_client_config" "test" {}
-
-resource "azurerm_resource_group" "test" {
-  name     = "acctestRG-mssql-%[1]d"
-  location = "%[2]s"
-}
 
 resource "azurerm_user_assigned_identity" "test" {
   resource_group_name = azurerm_resource_group.test.name
@@ -846,7 +781,7 @@ resource "azurerm_user_assigned_identity" "test" {
 }
 
 resource "azurerm_mssql_server" "test" {
-  name                         = "acctestsqlserver%[1]d"
+  name                         = "acctestsqlserver%[2]d"
   resource_group_name          = azurerm_resource_group.test.name
   location                     = azurerm_resource_group.test.location
   version                      = "12.0"
@@ -867,21 +802,14 @@ resource "azurerm_mssql_server" "test" {
 
   primary_user_assigned_identity_id = azurerm_user_assigned_identity.test.id
 }
-`, data.RandomInteger, data.Locations.Primary, enableAzureadAuthenticationOnly)
+`, r.template(data), data.RandomInteger, enableAzureadAuthenticationOnly)
 }
 
-func (MsSqlServerResource) tdeCMKServerDeployment(data acceptance.TestData) string {
+func (r MsSqlServerResource) tdeCMKServerDeployment(data acceptance.TestData) string {
 	return fmt.Sprintf(`
-provider "azurerm" {
-  features {}
-}
+%s
 
 data "azurerm_client_config" "test" {}
-
-resource "azurerm_resource_group" "test" {
-  name     = "acctestRG-mssql-%[1]d"
-  location = "%[2]s"
-}
 
 resource "azurerm_user_assigned_identity" "test" {
   resource_group_name = azurerm_resource_group.test.name
@@ -890,7 +818,7 @@ resource "azurerm_user_assigned_identity" "test" {
 }
 
 resource "azurerm_mssql_server" "test" {
-  name                         = "acctestsqlserver%[1]d"
+  name                         = "acctestsqlserver%[2]d"
   resource_group_name          = azurerm_resource_group.test.name
   location                     = azurerm_resource_group.test.location
   version                      = "12.0"
@@ -913,7 +841,7 @@ resource "azurerm_mssql_server" "test" {
 }
 
 resource "azurerm_key_vault" "test" {
-  name                        = "vault%[1]d"
+  name                        = "vault%[2]d"
   location                    = azurerm_resource_group.test.location
   resource_group_name         = azurerm_resource_group.test.name
   enabled_for_disk_encryption = true
@@ -948,21 +876,14 @@ resource "azurerm_key_vault_key" "test" {
 
   key_opts = ["unwrapKey", "wrapKey"]
 }
-`, data.RandomInteger, data.Locations.Primary, data.RandomString)
+`, r.template(data), data.RandomInteger, data.RandomString)
 }
 
-func (MsSqlServerResource) CMKServerTags(data acceptance.TestData, tag string) string {
+func (r MsSqlServerResource) CMKServerTags(data acceptance.TestData, tag string) string {
 	return fmt.Sprintf(`
-provider "azurerm" {
-  features {}
-}
+%s
 
 data "azurerm_client_config" "test" {}
-
-resource "azurerm_resource_group" "test" {
-  name     = "acctestRG-mssql-%[1]d"
-  location = "%[2]s"
-}
 
 resource "azurerm_user_assigned_identity" "test" {
   resource_group_name = azurerm_resource_group.test.name
@@ -971,7 +892,7 @@ resource "azurerm_user_assigned_identity" "test" {
 }
 
 resource "azurerm_mssql_server" "test" {
-  name                         = "acctestsqlserver%[1]d"
+  name                         = "acctestsqlserver%[2]d"
   resource_group_name          = azurerm_resource_group.test.name
   location                     = azurerm_resource_group.test.location
   version                      = "12.0"
@@ -998,7 +919,7 @@ resource "azurerm_mssql_server" "test" {
 }
 
 resource "azurerm_key_vault" "test" {
-  name                        = "vault%[1]d"
+  name                        = "vault%[2]d"
   location                    = azurerm_resource_group.test.location
   resource_group_name         = azurerm_resource_group.test.name
   enabled_for_disk_encryption = true
@@ -1033,21 +954,14 @@ resource "azurerm_key_vault_key" "test" {
 
   key_opts = ["unwrapKey", "wrapKey"]
 }
-`, data.RandomInteger, data.Locations.Primary, data.RandomString, tag)
+`, r.template(data), data.RandomInteger, data.RandomString, tag)
 }
 
-func (MsSqlServerResource) CMKServerNoTags(data acceptance.TestData) string {
+func (r MsSqlServerResource) CMKServerNoTags(data acceptance.TestData) string {
 	return fmt.Sprintf(`
-provider "azurerm" {
-  features {}
-}
+%s
 
 data "azurerm_client_config" "test" {}
-
-resource "azurerm_resource_group" "test" {
-  name     = "acctestRG-mssql-%[1]d"
-  location = "%[2]s"
-}
 
 resource "azurerm_user_assigned_identity" "test" {
   resource_group_name = azurerm_resource_group.test.name
@@ -1056,7 +970,7 @@ resource "azurerm_user_assigned_identity" "test" {
 }
 
 resource "azurerm_mssql_server" "test" {
-  name                         = "acctestsqlserver%[1]d"
+  name                         = "acctestsqlserver%[2]d"
   resource_group_name          = azurerm_resource_group.test.name
   location                     = azurerm_resource_group.test.location
   version                      = "12.0"
@@ -1079,7 +993,7 @@ resource "azurerm_mssql_server" "test" {
 }
 
 resource "azurerm_key_vault" "test" {
-  name                        = "vault%[1]d"
+  name                        = "vault%[2]d"
   location                    = azurerm_resource_group.test.location
   resource_group_name         = azurerm_resource_group.test.name
   enabled_for_disk_encryption = true
@@ -1114,10 +1028,30 @@ resource "azurerm_key_vault_key" "test" {
 
   key_opts = ["unwrapKey", "wrapKey"]
 }
-`, data.RandomInteger, data.Locations.Primary, data.RandomString)
+`, r.template(data), data.RandomInteger, data.RandomString)
 }
 
-func (MsSqlServerResource) writeOnlyAdminLoginPassword(data acceptance.TestData, secret string, version int) string {
+func (r MsSqlServerResource) writeOnlyAdminLoginPassword(data acceptance.TestData, secret string, version int) string {
+	return fmt.Sprintf(`
+%[1]s
+
+%[2]s
+
+resource "azurerm_mssql_server" "test" {
+  name                                    = "acctestsqlserver%[3]d"
+  resource_group_name                     = azurerm_resource_group.test.name
+  location                                = azurerm_resource_group.test.location
+  version                                 = "12.0"
+  administrator_login                     = "missadministrator-${random_string.test.result}"
+  administrator_login_password_wo_version = %[4]d
+  administrator_login_password_wo         = ephemeral.azurerm_key_vault_secret.test.value
+
+  outbound_network_restriction_enabled = true
+}
+`, r.template(data), acceptance.WriteOnlyKeyVaultSecretTemplate(data, secret), data.RandomInteger, version)
+}
+
+func (MsSqlServerResource) template(data acceptance.TestData) string {
 	return fmt.Sprintf(`
 provider "azurerm" {
   features {}
@@ -1128,24 +1062,10 @@ resource "azurerm_resource_group" "test" {
   location = "%[2]s"
 }
 
-%s
-
-resource "random_string" "suffix" {
+resource "random_string" "test" {
   length  = 3
   special = false
   upper   = false
 }
-
-resource "azurerm_mssql_server" "test" {
-  name                                    = "acctestsqlserver%[1]d"
-  resource_group_name                     = azurerm_resource_group.test.name
-  location                                = azurerm_resource_group.test.location
-  version                                 = "12.0"
-  administrator_login                     = "missadministrator-${random_string.suffix.result}"
-  administrator_login_password_wo_version = %[4]d
-  administrator_login_password_wo         = ephemeral.azurerm_key_vault_secret.test.value
-
-  outbound_network_restriction_enabled = true
-}
-`, data.RandomInteger, data.Locations.Primary, acceptance.WriteOnlyKeyVaultSecretTemplate(data, secret), version)
+`, data.RandomInteger, data.Locations.Primary)
 }

--- a/internal/services/mssql/mssql_server_resource_test.go
+++ b/internal/services/mssql/mssql_server_resource_test.go
@@ -1130,12 +1130,18 @@ resource "azurerm_resource_group" "test" {
 
 %s
 
+resource "random_string" "suffix" {
+  length  = 3
+  special = false
+  upper   = false
+}
+
 resource "azurerm_mssql_server" "test" {
   name                                    = "acctestsqlserver%[1]d"
   resource_group_name                     = azurerm_resource_group.test.name
   location                                = azurerm_resource_group.test.location
   version                                 = "12.0"
-  administrator_login                     = "missadministrator"
+  administrator_login                     = "missadministrator-${random_string.suffix.result}"
   administrator_login_password_wo_version = %[4]d
   administrator_login_password_wo         = ephemeral.azurerm_key_vault_secret.test.value
 

